Add option to limit number of SMs used by matmul kernels

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144974

Newer matmul kernels, e.g. those targeting Hopper GPUs, sometime use a "persistent" schedule which consists in launching as many CUDA blocks as there are SMs on the GPU, with each such block then working on multiple output tiles in a row. This allows to eliminate the overhead of starting and finishing each tile, effectively doing cross-tile pipelining. In previous generations these latencies could be hidden by having multiple CUDA blocks per SM but, with blocks becoming larger, only one can run at a time per SM and thus this needs to be taken care of in software.

Persistent kernels become an issue when other kernels are running concurrently. The classical example is a NCCL communication kernel running in the background. In such cases the matmul expects to be able to use all the SMs but is prevented from doing so because some of the are busy. This can lead to its blocks being scheduled as two separate waves on the available SMs. This "wave quantization" can double the latency of the matmul kernels.

While we wait for smarter solutions, such as automatic load balancing among the blocks, an easy way to unblock ourselves is to tell the matmuls to only use a subset of the GPU's SMs. For this, I am introducing a global `sm_carveout` flag which can be used to specify how many SMs should be left available for other kernels.

For now I only change the cuBLAS kernels and the scaled-mm CUTLASS kernel. More kernels can be opted-in later.

I tested this change manually, by using the Kineto profiler to look up the grid size of a scaled-mm kernel with different values of `sm_carveout`, and making sure it changed. Suggestions are welcome for a more automated test.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

lw committed:
Update

[ghstack-poisoned]

albanD reviewed (commented):
Thanks for sending a proposal here!
My quick take on the feature is:
- We can't really experiment this out of core, so I'm willing to do some API experimentation here BUT we need to make sure all of this is private and easy to remove if needed.
- I'm not sure if a % reduction vs flat reduction is the best
- I am very curious if the context-manager from user-code going to work in practice. Another option might be for NCCL calls to set this setting? But with async execution, not sure how it's going to be behave in practice.

For this particular PR to be merged:
- make things private + warning for maybe future deletion
- Add a comment block next to the context definition explaining what this feature is, what it does, where it is read and the BC promises about it.
- tests
- move the python bindings to pybind

yifuwang commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -364,6 +364,7 @@ inline void bgemm_internal_cublaslt(CUDABLAS_BGEMM_ARGTYPES(Dtype)) {
   CuBlasLtMatmulDescriptor computeDesc(computeType, scaleType);
   computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_TRANSA, opa);
   computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_TRANSB, opb);
+  computeDesc.setAttribute<int32_t>(CUBLASLT_MATMUL_DESC_SM_COUNT_TARGET, at::cuda::getCurrentDeviceProperties()->multiProcessorCount - at::globalContext().SMCarveout());
```
Somewhat related question: I noticed that when this is unspecified, cuBLAS often does not utilize all the available SMs (it's very common to see 110–120 out of 132 SMs being utilized on H100, maybe except for the stream-k kernels).

- Do you know if the 'under-utilization' was actually for better matmul performance, or was cuBLAS trying to be carve-out conscious?
- Specifying a zero carve-out here means setting CUBLASLT_MATMUL_DESC_SM_COUNT_TARGET to 132 on H100. Would this change the SM utilization of those kernels that would use less than 132 SMs previously?

eqy replied:
cuBLAS may use thread block cluster sizes that reduce occupancy as an optimization, so it should have nothing to do with carveout if there was no option passed to carveout

"Using larger cluster sizes may reduce the maximum number of active blocks across the GPU (refer to [Occupancy](https://docs.nvidia.com/cuda/hopper-tuning-guide/index.html#sm-occupancy))."
see note on thread block clusters: https://docs.nvidia.com/cuda/hopper-tuning-guide/index.html#thread-block-clusters

eqy commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -364,6 +364,7 @@ inline void bgemm_internal_cublaslt(CUDABLAS_BGEMM_ARGTYPES(Dtype)) {
   CuBlasLtMatmulDescriptor computeDesc(computeType, scaleType);
   computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_TRANSA, opa);
   computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_TRANSB, opb);
+  computeDesc.setAttribute<int32_t>(CUBLASLT_MATMUL_DESC_SM_COUNT_TARGET, at::cuda::getCurrentDeviceProperties()->multiProcessorCount - at::globalContext().SMCarveout());
```
Maybe only call the API if carveout is specified ? (seems unlikely in the average user case?)

eqy reviewed (commented):
please add a test to `test_cuda.py` similar to other CUDA context functions

eqy commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -1248,6 +1249,7 @@ void gemm_and_bias(
   computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_TRANSA, transa);
   cublasOperation_t transb = transpose_mat2 ? CUBLAS_OP_T : CUBLAS_OP_N;
   computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_TRANSB, transb);
+  computeDesc.setAttribute<int32_t>(CUBLASLT_MATMUL_DESC_SM_COUNT_TARGET, at::cuda::getCurrentDeviceProperties()->multiProcessorCount - at::globalContext().SMCarveout());
```
Looks like these calls need ifdefs as apparently ROCM does not have the same API

eellison commented:
cc @aakhundov @davidberard98 we should respect this in the inductor persistent kernels

davidberard98 commented:
@eellison filed https://github.com/pytorch/pytorch/issues/145115 to track inductor persistent kernel support for this

lw committed:
Update

[ghstack-poisoned]

lw commented:
- I changed the names of the Context methods to mark them private (leading underscore) and experimental
- I also added a comment with context
- And I logged a warning the first time the carveout is set
- I changed the Python bindings to use pybind
- I removed the `torch.cuda.matmul.sm_carvout` attribute, forcing users to go through the `torch._C._set_sm_carvout` method, so it's clearer it's private/experimental
- I made the value become optional, so that the default of None ensures we don't set anything at all and reproduce the current behavior exactly
- I disabled this on ROCm

As for "absolute" vs "relative", I went for an absolute _negative_ value because, usually, NCCL uses a fixed amount of SMs (e.g., 16 or 24) and that value can be easily controlled with `NCCL_MAX_NCHANNELS`. Thus I expect users to mainly want to tell matmuls to leave available a specific amount of SMs. Hence the "carveout".

As for testing, I have currently no idea how to reliably test this, because (at least for cuBLAS) the actual algorithm used is an implementation detail thus we cannot really assert that by setting this value we correctly triggered cuBLAS to avoid persistent kernels or to launch fewer blocks. This may be easier to do with our CUTLASS kernels but, even there, we'd likely need to use Kineto to profile the kernels to make sure they use the number of blocks we expect, and I couldn't find how to do so. Help is appreciated! :)

lw commented on torch/_C/__init__.pyi.in:
```diff
@@ -1210,6 +1210,8 @@ def _get_cublas_allow_bf16_reduced_precision_reduction() -> _bool: ...  # THPMod
 def _set_cublas_allow_bf16_reduced_precision_reduction(
     arg: _bool,
 ) -> None: ...  # THPModule_setAllowBF16ReductionCuBLAS
+def _get_sm_carveout() -> _int: ...  # THPModule_SMCarveout
+def _set_sm_carveout(arg: _int) -> None: ...  # THPModule_setSMCarveout
```
TODO: Remove these comments as they don't apply anymore

lw committed:
Update

[ghstack-poisoned]

albanD commented on aten/src/ATen/Context.cpp:
```diff
@@ -411,6 +411,19 @@ void Context::setAllowBF16ReductionCuBLAS(bool b) {
+  return sm_carveout;
+}
+
+void Context::_setSMCarveout_EXPERIMENTAL(std::optional<int> c) {
```
please use int32_t or int64_t here for fixed bitwidth.

albanD reviewed (approved):
Thanks for the cleanup!
Let's see how the experimentation for this goes.

Only small nit for int bitwidth but good to go otherwise

lw committed:
Update

[ghstack-poisoned]

lw committed:
Update

[ghstack-poisoned]

lw commented:
@pytorchbot merge

lw committed:
Update

[ghstack-poisoned]

lw committed:
Update

[ghstack-poisoned]

lw commented:
@pytorchbot merge -f "Rebased to fix trivial merge conflict, CI was all green before"

wdvr commented:
@pytorchmergebot revert -m 'reverting in order to revert #147548 that causes a merge conflict' -c ghfirst

wdvr commented:
@pytorchmergebot rebase
